src/aave-v3/SupplyVault.sol

```diff
@@ -217,36 +196,61 @@ contract SupplyVault is ISupplyVault, SupplyVaultBase {
             (rewardTokens, claimedAmounts) = morpho.claimRewards(poolTokens, false);
         }
 
-        uint256 supply = totalSupply();
         for (uint256 i; i < rewardTokens.length; ++i) {
             address rewardToken = rewardTokens[i];
             uint256 claimedAmount = claimedAmounts[i];
+            UserRewardsData storage userRewardsData = userRewards[rewardToken][_user];
             uint128 rewardsIndexMem;
 
-            if (supply > 0 && claimedAmount > 0) {
+            if (totalSupply() > 0 && claimedAmount > 0) {
                 rewardsIndexMem =
                     rewardsIndex[rewardToken] +
-                    claimedAmount.mulDivDown(SCALE, supply).safeCastTo128();
+                    _getUnaccruedRewardIndex(claimedAmount);
                 rewardsIndex[rewardToken] = rewardsIndexMem;
             } else rewardsIndexMem = rewardsIndex[rewardToken];
 
-            UserRewardsData storage userRewardsData = userRewards[rewardToken][_user];
-            uint256 rewardsIndexDiff;
-
-            // Safe because we always have `rewardsIndex` >= `userRewardsData.index`.
-            unchecked {
-                rewardsIndexDiff = rewardsIndexMem - userRewardsData.index;
-            }
-
-            if (rewardsIndexDiff > 0) {
-                uint128 accruedRewards = balanceOf(_user)
-                .mulDivDown(rewardsIndexDiff, SCALE)
-                .safeCastTo128();
-                userRewardsData.unclaimed += accruedRewards;
+            if (rewardsIndexMem > userRewardsData.index) {
+                uint128 accruedReward = _getUnaccruedRewardFromUnaccruedRewardIndex(
+                    _user,
+                    rewardsIndexMem - userRewardsData.index
+                );
+                userRewardsData.unclaimed += accruedReward;
                 userRewardsData.index = rewardsIndexMem;
 
-                emit Accrued(rewardToken, _user, rewardsIndexMem, accruedRewards);
+                emit Accrued(rewardToken, _user, rewardsIndexMem, accruedReward);
             }
         }
     }
+
+    function _getUnaccruedReward(
+        address _user,
+        address _rewardToken,
+        uint256 _claimableReward
+    ) internal view returns (uint128 unaccruedReward) {
+        unaccruedReward = _getUnaccruedRewardFromUnaccruedRewardIndex(
+            _user,
+            _getUnaccruedRewardIndex(_claimableReward) + // The unaccrued reward index
+                rewardsIndex[_rewardToken] -
+                userRewards[_rewardToken][_user].index // The difference between the current reward index and the user's index
+        );
+    }
+
+    function _getUnaccruedRewardFromUnaccruedRewardIndex(
```

Comment:
Ok so this `_unaccruedRewardIndex` name is misleading since right above you have:
```solidity
unaccruedReward = _getUnaccruedRewardFromUnaccruedRewardIndex(
   balanceOf(_user),
    _getUnaccruedRewardIndex(_claimableReward, _totalSupply) + // The unaccrued reward index
          rewardsIndex[_rewardToken] -
           userRewards[_rewardToken][_user].index // The difference between the current reward index and the 
   user's index
);
```
`// The unaccrued reward index` this comment is exactly the input variable.

Reply:
- `_getUnaccruedRewardsFromRewardsIndexAccrual`
- `_getUnaccruedRewardsFromRewardsIndexIncrease`
- `_getUnaccruedRewardsFromRewardsIndexIncrement`